### ISSUE TITLE
Not able to fetch Audiobook URL

### ISSUE THREAD

**Issue by JaegerMaster** [jdalbey/libbyfetch#23]
Script is timing out after waiting to fetch the Audiobook URL. Happening with all books and multiple libraries

`Initializing webdriver.
Reading library card configuration file.
Loading library page for 'wccls'
Starting signin to Washington County Cooperative Library Services.
Library card number entered.
PIN entered.
Sign In completed.
Loans page loaded.
1. Beneath Stone and Sacri                                                      
Enter the number of the desired title: 1
Opening Audiobook.
Retrieving audiobook url...
waiting
waiting
waiting
waiting
waiting
waiting
waiting
waiting
waiting
waiting
Timed out waiting for audiobook url.
Page says: An error was encountered while opening this audiobook.`

**Comment by jdalbey** [comment]
Oh no, I'm sorry this is happening for you.   I discovered this problem and reported it as issue #17 so you can read about it there.  It's possible that Libby is blocking users who are downloading MP3's with libbyfetch.  (See 'Warning' in README).  If this is true that would be very bad. I have installed a fix 57c364b418224680316bf4113c581bddd43f26c5 so that it won't timeout and instead displays an error.  But that doesn't repair the underlying issue, which is that Libby is not allowing your card to access materials.  Try to manually play an audio book and see what happens.

**Comment by jdalbey** [comment]
Duplicate of #17

**Comment by JaegerMaster** [comment]
Hi @jdalbey sorry for opening this issue as this started happening suddenly and I am using it with v1.04 so don't think it is related to #17.

**Comment by jdalbey** [comment]
@JaegerMaster I'm sorry for the confusion.  The behavior you reported above, where libbyfetch prints "waiting" ten times and then says "Time out" is what results from the fix I mentioned above (https://github.com/jdalbey/libbyfetch/commit/57c364b418224680316bf4113c581bddd43f26c5).  I see now that the message displayed doesn't provide very much troubleshooting information. I can add a better explanation in the future. 

To see what is causing the timeout, I suggest that you manually login to libbyapp.com and try to play an audiobook you have on loan.  I'm guessing that it won't allow you to access the audiobook. If that's the case then libbyfetch won't work for you either.

**Comment by JaegerMaster** [comment]
@jdalbey I am able able to play the audiobook on libbyapp.

Also, this is happening on multiple cards, tried it on multiple devices to rule out IP blocks.

I am not sure if this is something about the card being blacklisted by Overdrive.

**Comment by JaegerMaster** [comment]
If needed, I can share the library card details to test this further.

**Comment by jdalbey** [comment]
@JaegerMaster Thanks for the additional info.  I see you are able to play the audiobook on libbyapp. So that sounds like your issue is slightly different than the one previously reported so I'm going to reopen this issue.

**Comment by jdalbey** [comment]
@JaegerMaster You said that "this started happening suddenly" so you were able to download MP3 files with previous versions of libbyfetch?  Did the error occur when you updated to the current version?

**Comment by JaegerMaster** [comment]
Hi, I was able to download with 1.04 as well. Worked with 2-3 books. Then suddenly started happening with the 4th book and now happens with everything. I tried 2 different cards. Will try with a 3 card and let you know.
Will mail you the card number.

**Comment by jdalbey** [comment]
That's odd that it worked for a few books then stopped.  Quite a head scratcher.

**Comment by JaegerMaster** [comment]
Not working with the 3rd card as well. Don't think this is an IP block as didn't work on a different machine as well. If you can add a more detailed log option, I can help with the logs.

**Comment by jdalbey** [comment]
@JaegerMaster  Thanks, at last I was able to reproduce the error.  I've gathered these two examples for diagnosis:

1. Here's the developer [console log](https://gist.github.com/jdalbey/09bf8a0cef3e77d5a1ba2962e3596cb9) from a successful audiobook load after logging in manually (Firefox).

2. Here's a [log of HTTP Requests](https://gist.github.com/jdalbey/da23b135730f37f5ba5db47bf542312d) captured by libbyfetch form a run that produced "An error was encountered ..." message trying to open an audiobook. 

By studying these and looking for discrepancies I hope to identify why the manual process works but libbyfetch fails.

**Comment by JaegerMaster** [comment]
Thanks @jdalbey for looking into this.

**Comment by flyswimr** [comment]
Were the books that were tested already checked out and accessed at least one time prior via libbyapp.com or Overdrive, and prior to trying any of the LF scripts?  I seem to recall I had the same thing going on until I checked out a new audio book and tried to access via libbyapp.com.  At that point could not listen via the browser or access via LF.

**Comment by jdalbey** [comment]
@flyswimr Very interesting.  To clarify, you were able to listen to audiobooks via libbyapp.com, and you tried Libby Fetch and it worked okay?  But after checking out a new audiobook, you were no longer able to access any loans via libbyapp.com?

**Comment by JaegerMaster** [comment]
Yes. Tried playing the book on libbyapp and then downloading using the script. Didn't work.

**Comment by lsixecho** [comment]
I disabled headless mode so I could see where it failed and it displays the audiobook book but underneath says "An error occurred while opening this audiobook."
<br />
<img width="468" alt="Screenshot 2024-11-23 at 1 27 35 PM" src="https://github.com/user-attachments/assets/b5ddecfc-16f0-458c-85b1-cde5495ec543">
<br />
This audiobook plays perfectly on a normal chrome window, just not the libbyfetch window.

**Comment by jdalbey** [comment]
@lsixecho Thanks for confirming this behavior. This seems like the same error others are experiencing. When in headless mode libbyfetch will display the error after attempting to retrieve the audiobook URL times out (as seen in the initial reporting of this issue). 
 Do you recall if you acquired the loan of this audiobook BEFORE you tried to download any other books with libbyfetch?  Or had you downloaded other audiobooks with libbyfetch and then subsequently acquired this one on your "shelf". 

I'm curious what will happen if you now acquire a new audiobook loan and try to play it manually (not with libbyfetch).  Does that work?  Are you able to checkout additional audiobooks and listen to them through the web browser?  If you have time to try it let me know what happens.

**Comment by lsixecho** [comment]
> @lsixecho Thanks for confirming this behavior. This seems like the same error others are experiencing. When in headless mode libbyfetch will display the error after attempting to retrieve the audiobook URL times out (as seen in the initial reporting of this issue). Do you recall if you acquired the loan of this audiobook BEFORE you tried to download any other books with libbyfetch? Or had you downloaded other audiobooks with libbyfetch and then subsequently acquired this one on your "shelf".
> 
> I'm curious what will happen if you now acquire a new audiobook loan and try to play it manually (not with libbyfetch). Does that work? Are you able to checkout additional audiobooks and listen to them through the web browser? If you have time to try it let me know what happens.

Ok, I tried another couple audiobooks on different cards and they all have the same waiting/timeout error. I tried playing first, then downloading first alternatively. They all play normally during a default session of chrome, regardless of being played first or download attempted first, so I am guessing it's a selenium issue. I have previously downloaded many audiobooks before this sudden error and none since. Any other ideas for debugging?

**Comment by jdalbey** [comment]
@lsixecho Thanks, that's a very interesting report.  It's good that you can still play audiobooks through the browser.  That is no longer working for me; even the browser gives me the "error encountered" message.   If you would like to help debug I could use some logs from using Libby in the browser.  Here are some directions for how to [Capture HTTP requests](https://telegra.ph/Markdown-demo-11-24) from the browser. If you could capture a complete session using Libby to sign in, go to your loans page, and play an audiobook.  Then post the resulting file containing HAR log as an attachment to a comment here.  That way I can compare a session that succeeds against one that fails and see what's happening.
Also if you could report what Settings you have enabled in Chrome regarding Cookie handling.

**Comment by jdalbey** [comment]
I found that one cause of "An error was encountered" message when using Libby in the browser is if you have ad blocker or privacy extensions enabled.   Try disabling ad blocker or similar privacy extensions while using libbyapp.com.

**Comment by OGQueenie** [comment]
I too am experiencing this issue, I can access the books just fine via the app or browser, I have yet to successfully download a book via your program though

**Comment by ApKNJ** [comment]
I'm seeing this issue as well.  I'm testing on a new Linux VM set up just for this purpose with a fresh install of Chromium, no extensions.  Issue occurs on a book that was already on loan before trying lf for the first time, and one that was borrowed after trying lf, no difference. Both books will play in Chromium normally.

**Comment by lsixecho** [comment]
> @lsixecho Thanks, that's a very interesting report. It's good that you can still play audiobooks through the browser. That is no longer working for me; even the browser gives me the "error encountered" message. If you would like to help debug I could use some logs from using Libby in the browser. Here are some directions for how to [Capture HTTP requests](https://telegra.ph/Markdown-demo-11-24) from the browser. If you could capture a complete session using Libby to sign in, go to your loans page, and play an audiobook. Then post the resulting file containing HAR log as an attachment to a comment here. That way I can compare a session that succeeds against one that fails and see what's happening. Also if you could report what Settings you have enabled in Chrome regarding Cookie handling.

Here are the HAR files. I'm not sure what you mean but cookie settings, presumably just the default, but they certainly haven't changed and downloading was working before this error.

[Har Files.zip](https://github.com/user-attachments/files/17905920/Har.Files.zip)

Not-working.har is the libbyfetch session (logs turned on during pause to select audiobook, which is the first change you get)

Working.har is normal chrome session that plays audiobook (the full capture, although libbyfetch doesn't fail until the play stage anyway)

**Comment by jdalbey** [comment]
@lsixecho  Thanks for taking time to create these HAR files.  I'm looking into this problem but haven't yet isolated what's going wrong.

**Comment by lsixecho** [comment]
> @lsixecho Thanks for taking time to create these HAR files. I'm looking into this problem but haven't yet isolated what's going wrong.

I did a lot more testing today with various methods, and I do think they are blocking selenium efforts, to the point that I can no longer even play the audiobook under any circumstances, whereas I was able to play it in the default chrome profile nearly all day. I tried different ip addresses, incognito, tor browser, VPN, different computer on a different ISP, etc and in no method am I able to play it at all anymore, with a new error message "_**Access to this magazine is not allowed with your active library card.**_" There may be a cooldown period and I have 14 days left with this particular audiobook, so I can try again near the end of that period. 

I was attempting to diagnose why a default session would behave differently than a debugging/selenium session. I also tried undetected_chromedriver, all with no success. I really feel that this is the key to solving it. What are they doing to detect the non-user session and how to mask that enough to access the page without it erroring out? I know it's cat and mouse, but the cat is winning right now, and I'm trying to be a good mouse.

Perhaps if this method is stalling out, we could automate the gathering of the mp3 urls and save them to a text file, then automate their download with python. I suspect that if I had a clean card and/or a clean loan, I could step through the audiobook, gather the mp3 locations, and download them, just like your script does, but without the headless selenium automation so the book loads properly. It would be much more manual than the current script, but the failure point is the opening of the book, so if we have to clear that hurdle as a human, at least the tedium of mp3 url gathering, downloading, and renaming could be automated. 

Still works with the iPhone app though. 😔

**Comment by lsixecho** [comment]
And one more follow up is that my library still has it's overdrive page active (even thought downloads are no longer available in the .odm format). And I can play the audiobook with no restrictions here, performing the same method of inspect tab, chapter skipping, looking for mp3 media, now in the https://odrmediaclips.cachefly.net/ url. I wonder if we could modify the script to use the overdrive site to pull the mp3s, until they shut it down completely.

**Comment by jdalbey** [comment]
@lsixecho Thanks for this detailed report.  You have put a lot of effort into this.  I applaud your persistence. I'm sorry that your library card is now blocked from playing audiobooks.  I'm getting that same message about "Access to your magazine ..." which is weird because the loan isn't a magazine.  I'm not so sure they are blocking selenium specifically as it is that the program is failing some kind of authenticaion and after some number of attempts the server blocks that card.  
One additional data point: I recorded a script in Selenium IDE that when played, successfully opens the Audiobook player.  But when I export that script to Python it gets an error when attempting to open the player.
Yes, the crux of the technical problem is gathering the URL's.  When I initially coded LibbyFetch the URL's were all sequential (Part-01, Part-02, ...).  But they have recently changed it to add some encrypted string the query parameters to you can't just increment the integer value.  Now you have to gather all the URL's, not just the first one.

**Comment by flyswimr** [comment]
> @flyswimr Very interesting. To clarify, you were able to listen to audiobooks via libbyapp.com, and you tried Libby Fetch and it worked okay? But after checking out a new audiobook, you were no longer able to access any loans via libbyapp.com?

True, my recollection is that they left my existing checkouts alone but any new checkouts received that message about not being able to 'access that magazine'.   I'm curious if this is a permanent shadow ban or if it is removed after x days.

**Comment by kailashkalyani** [comment]
Hi, it appears that libbyapp does detect if an automation is being used. I tried it using chromium with and without the webdriver - with the webdriver, I get a similar error and without the webdriver, it appears to be working fine. I tried replacing the selenium webdriver with undetected_chromedriver, but that did not resolve the issue. Sorry, looks like a better option might be to explore using a chrome plug-in.

**Comment by flyswimr** [comment]
Is it accurate then that they are now detecting automation with LF, and they weren't a couple of weeks ago?  It also seems that Libby then bans access to materials.  They still allow materials to be checked out using the libby app.  

I want to say I remember seeing a chrome plug-in developed but it was abandoned, ostensibly due to Overdrive/Libby detecting that individual mp3 were being accessed due to skippping chapters or something (but also downloading individual mp3s).

**Comment by kailashkalyani** [comment]
I guess you mean this one here: https://github.com/bookbonobo/libby-download-extension - yeah I just tested it. It stopped working about a year ago. Earlier while testing, as I skipped chapters and switched between books, they blocked access to one title for a few seconds. So, there is likely some session monitoring underway as well.

**Comment by flyswimr** [comment]
@kailashkalyani Correct.  What is interesting is I was able to use the browser console in Firefox from my library's Overdrive site to locate the cachefly links, open them individually (let them play short time), and then do a save as for each.  And so far not blocked, and why should that happen, they are valid links and the browser allows this.  Am assuming this is similar to what Libby fetch does and session monitoring is in play.  And if developers at Libby are listening - please give a way to play content on smaller devices, mp3 player, Sandisk, etc.  You all might even save some time redesigning your work, and if so, that means there is a demand for the capability.

**Comment by OGQueenie** [comment]
Yeah it's definitely seeming like there won't be an easy/automated way to download books from Libby, with Overdrive being retired it looks like we're gonna have to look to alternative methods to get mp3s of audiobooks unfortunately

**Comment by kailashkalyani** [comment]
Frankly, it all depends on whether they are using sophisticated pattern
detection systems to analyse session behaviors. If so, then there is not a
lot one can do other than playing the whole audiobook and capturing the
media. And I am inclined to think they do. Look at the tracking available
to you on what segment you listened to and if you skipped forward or
rewound in an audiobook.
https://www.fullstory.com/blog/user-sessions/

**Comment by AnonJohn** [comment]
Is it possible to take a different approach: Have the computer listen to the whole book, in real time? That is, use libbyfetch's chapter parsing, but actually play the audio chapter by chapter? Then grab the audio stream at the OS level or browser level and re-encode mp3s? 

For e-books (that I'll only listen to once) I'm not too concerned about quality degradation. And I can wait 10-20 hours for it to happen in the background. Thoughts?

**Comment by kailashkalyani** [comment]
It should actually be easier. In effect you could "mute" your tab, choose
playback speed 2x, hit play and monitor the network accesses - each time a
new cachefly link is made available, download that file. The approach least
likely to be "detected" would be via plugin, I think.

**Comment by raybies808** [comment]
Since libbyfetch has stopped working, I've been using the Java script from here: https://remedylegacy.com/Files/LibbyDownload-3.2.zip. This requires you to install Java, open Libby in your browser, go to the audiobook you want to save, open the Developer Tools in your browser, play the audiobook then fast forward all the way to the end, then save the session to a HAR file. The script will then read the HAR file and save the audiobook mp3 files. It takes more work but still works, the step by step details are in a readme file within the zip.

**Comment by flyswimr** [comment]
I checked to see if LibbyDownload would work with LibbyDownload using Overdrive w/s (for those that are blocked from using Libbyapp.com).  It doesn't because it is looking for a table of contents.   For now can filter to the individual mp3 files.  Good enough for now.

**Comment by schuchuva** [comment]
I use this Firefox/Chrome/Edge extension:
https://www.downloadhelper.net/
I have to start playing every file from Ovedrive book's contents page and then download mp3 file.
Inconvenient but works so far.

**Comment by lsixecho** [comment]
> I use this Firefox/Chrome/Edge extension: https://www.downloadhelper.net/ I have to start playing every file from Ovedrive book's contents page and then download mp3 file. Inconvenient but works so far.

This is absolutely the easiest method so far. I used this extension on Overdrive play (since I am now blocked, seemingly permanently on libby). I would presume it would work the same for Libby, but can't test it. 

It's a dead simple solution, easier than getting python working for most people. I assume it's doing what we discussed about grabbing MP3 links from the network tab and displaying in the GUI with a download button. 

All I had to do was clean up the extras (for some reasons there were two of everything expect the last chapter/section) and use my preferred app to join the files (which you wouldn't have to do for an mp3 player if you didn't want to).

For me (or until this is further restricted) I will consider this closed with libbyfetch dead, but I am so glad to have alternative now that honestly took me about as much time to run as libbyfetch. 

Thank you so much @schuchuva and the team here!

**Comment by jdalbey** [comment]
Regarding @raybies808 suggestion about the Java downloader, I took a look at their code and I see that not only does it grab the URL from the HAR file, it also grabs all the request headers.  So when it requests the .mp3 file it can duplicate all the request headers that libbyapp uses. 
Also, there's a quicker way than suggested in their README.  It isn't necessary to use the 15sec fast forward.  You can simple click on each chapter in the table of contents in sequence. It's much quicker than fast forwarding through the entire book.

**Comment by flyswimr** [comment]
@isixecho, @schuchuva It does work pretty well on Overdrive, thank you!  It would be interesting to see if it works on Libbyapp.com without triggering adverse consequences.

**Comment by lsixecho** [comment]
I made a little tutorial, because I found that downloadhelper extension missed a bunch of files for some unknown reason. I'm not able to test on libby because I am banned because of libbyfetch. I realized that for 4/5 books I tried, downloadhelper missed several mp3s and I had to go back and do it manually anyway, so I made a little tutorial for how I eventually was successful, but it's an ugly tutorial because I threw it together quick and had to blur/crop personal stuff. 

It's an uglier method than my last post's praise would seem to indicate, but I'm just happy something works at all!


https://github.com/user-attachments/assets/26af8842-bf3c-40b3-b3db-595b2b7ad0d2

**Comment by flyswimr** [comment]
That's good.  I've had the same thing happen when using the provided control on the right hand side that advances "x"m (minutes).  Holding the right arrow on the keyboard down advances 15 seconds at a time but once it goes into repeat mode goes by rapidly enough (and so far haven't missed a file).

**Comment by schuchuva** [comment]
I verified and confirm that downloadhelper extension works on libbyapp.com.

@lsixecho - I'm curious how you managed to miss some files with downloadhelper extension. I have another problem - links on overdrive/libbyapp contents page often point to same file - I have to guess and verify link to next file.

**Comment by lsixecho** [comment]
> I verified and confirm that downloadhelper extension works on libbyapp.com.
> 
> @lsixecho - I'm curious how you managed to miss some files with downloadhelper extension. I have another problem - links on overdrive/libbyapp contents page often point to same file - I have to guess and verify link to next file.

I'm not too sure. I had the 19 hour book "On Call" by Fauci to use as a test, and when I downloaded everything in the extension, it added up to only 12 hours. I even watched the helper (you can have it in the sidebar so you can watch new files pop-in) and I would skip to a hour-ish long file that I missed in the first pass, and it would just not see it for some reason. Like I said, this happened with 4/5 books, and only on one of them did it grab everything.

Not too sure if it's something I'm doing wrong, but I don't think so since it won't show as a "found" MP3, at least until I click on the network tabbed one to open it, then it can find it just fine.

**Comment by schuchuva** [comment]
I have this problem sometimes with downloadhelper extension - it stops detecting new files when clicking on content page links. I have to refresh page and then all the missing files appear in downloadhelper dropdown button.

**Comment by CometKingDoodle** [comment]
Thanks for everyone's input and comments.  I find that download helper is kind of messy, but overall it works.   To avoid downloading duplicate files this is what I do.  
1. Download the first file and note the duration.    
2. Look at the table of contents and start play at that duration and download the 2nd file.
3. Note the duration of the 2nd file and add it to the first duration.  I do this in a spreadsheet that looks like this:![2024-12-10_22-37-17](https://github.com/user-attachments/assets/83e21b32-3e9d-4635-9810-2697dd632f21)
4. The cumulative duration tracks really well with the durations on the table of contents:![2024-12-10_22-38-23](https://github.com/user-attachments/assets/0f940d0c-779e-4be8-856d-7bad999fa9c9)
Its a bit tedious but I end up with a set of files where the total duration matches what Overdrive says.  Would love any suggestions to streamline this process.

**Comment by manofthemountain** [comment]
one other way to keep track of what parts you have is to look at the GET requests, they have the part number they are requesting. 
![Screenshot 2024-12-11 at 11 16 02 AM](https://github.com/user-attachments/assets/7ebc6d14-bed8-40da-8930-7ed4f9a267c9)
